R/ds.boundedVarianceDP.R

```diff
@@ -7,18 +9,58 @@
 #' @param datasources a list of \code{\link{DSConnection-class}} 
 #' objects obtained after login. If the \code{datasources} argument is not specified
 #' the default set of connections will be used: see \code{\link{datashield.connections_default}}.
+#' @param type a character string that represents the type of analysis to carry out.
+#' This can be set as \code{'combine'}, \code{'split'} or \code{'both'}.
 #'
 #' @return \code{ds.boundedVarianceDP} returns a differentially private covariance
 #' @export
 
-ds.boundedVarianceDP <- function(input_data, epsilon, lower_bound, upper_bound, datasources=NULL) {
+ds.boundedVarianceDP <- function(input_data, epsilon, lower_bound, upper_bound, datasources=NULL, type="combine") {
 
   if (is.null(datasources)) {
     datasources <- DSI::datashield.connections_find()
   }
 
-  cally <- paste0("boundedVarianceDP(", input_data, ", ", epsilon, ", ", lower_bound, ", ", upper_bound, ")")
-  result <- DSI::datashield.aggregate(datasources, as.symbol(cally))
+  split_epsilon <- epsilon/3
+
+  SumSquares <- getAggregation(datasources, paste0("sumSquaresDP(", input_data, ", ", split_epsilon, ", ", lower_bound, ", ", upper_bound, ")"))
+  Sum <- getAggregation(datasources, paste0("sumDP(", input_data, ", ", split_epsilon, ", ", lower_bound, ", ", upper_bound, ")"))
+  Nvalid <- getAggregation(datasources, paste0("numValidDP(", input_data, ", ", split_epsilon, ")"))
+
+  Nstudies <- length(datasources)
 
-  return(result)
+  variance.combine <- computeVarCombine(Nstudies, Sum, SumSquares, Nvalid)
+  variance.split <- computeVarSplit(Nstudies, Sum, SumSquares, Nvalid)
+
+  if (type == "both") {
+      return(list(Variance.by.Study=variance.split,Global.Variance=variance.combine,Nstudies=Nstudies))
+  } else if (type == "combine") {
+      return(list(Global.Variance=variance.combine,Nstudies=Nstudies))
+  } else if (type == "split") {
+      return(list(Variance.by.Study=variance.split,Nstudies=Nstudies))
+  } else {
+      stop("type must be either 'both', 'combine' or 'split'")
+  }
```

Comment:
We could even check that at the beginning of the function to avoid useless computations (and budget expenses!)